Feature/co 685 get post patch delete event

jaredkosanovic opened the pull request.

Commit:
reformatting basic files

Commit:
added core classes for an event

Commit:
minor change to add bin as a directory to look in

Commit:
added files to invoke basic functionality

Commit:
removed skeleton class

Commit:
fixed types for testing

Commit:
Added instance functionality

Commit:
added groovy packages

Commit:
added class for custom field entry

Commit:
added instance and custom field functionality

Commit:
unused imports

Commit:
added class for filters

Commit:
updated event mapper for filters

Commit:
changed regex and moved instance dao call

Commit:
added get all json

Commit:
minor changes before adding new method

Commit:
fixes and added post

fixed issue with time formatting and time zone, changed name to eventID, added basic functionality for posting a new event including UUID generation and validation

Commit:
added eventid to attributes

Commit:
add response for conflict 409

Commit:
implemented more post features

implemented custom fields, filters, and instances

Commit:
minor changes

Commit:
changed date formatting

took advantage of java 8 and utilized java.time. No more fussing about with timezone offsets

Commit:
moved time manipulation methods

moved them to the instance mapper

Commit:
clean up the DAO

Commit:
exception handling and cleanup

added some exception handling and created a method for creating a result object

Commit:
added DELETE

added DELETE by id, which really just updates the db to soft delete an event

Commit:
beans, the magical fruit

Commit:
post with result object

per JSONapi, we should be posting and patching using full result objects

Commit:
post resultObject

Commit:
clean up

Commit:
had to create a patch interface

Commit:
removed eventid from attributes

Commit:
added basic PATCH

still need to add functionality to patch instances, custom fields, and filters

Commit:
PATCH fixes, added functionality

fixed PATCH issue of deleting values of attributes that were omitted from a request. Added functionality to update customfields, filters, and instances, and also add instances. Also implemented a rudementary DELETE instances function, whereby setting starttime and endtime to null soft deletes the instance. This is similar to how JSONapi specifies to delete a relationship, although with regards to relationships, this API isn't totally in line with JSONapi spec.

Commit:
add class for common error messages

Commit:
clean up

Commit:
add logging

Commit:
codenarc fix

Commit:
add more logic/steps for PATCHing instances

Commit:
updated readme

Commit:
clean up and comments

floppyears commented on src/main/groovy/edu/oregonstate/mist/api/Resource.groovy:
```diff
@@ -94,6 +94,20 @@ abstract class Resource {
     }
 
     /**
+     * Returns a builder for an HTTP 409 ("conflict") response with an error message as body.
```
Let's create a ticket so that we can add this functionality to the skeleton.

jaredkosanovic replied:
Created CO-700.

floppyears commented on src/main/groovy/edu/oregonstate/mist/api/PATCH.groovy:
```diff
@@ -0,0 +1,12 @@
+package edu.oregonstate.mist.api
```
This should be part of the skeleton. Let's create a ticket for it.

jaredkosanovic replied:
Created CO-701

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/db/EventsDAO.groovy:
```diff
@@ -0,0 +1,264 @@
+        VALUES (:id,
+            :title,
+            :description,
+            (SELECT PLACE_ID FROM PLACES
```
Can the place be deleted? If yes, should we check for null?

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/db/EventsDAO.groovy:
```diff
@@ -0,0 +1,264 @@
+            :description,
+            (SELECT PLACE_ID FROM PLACES
+                WHERE NAME = :location),
+            (SELECT GROUP_ID FROM GROUPS
```
Same as above

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/db/EventsDAO.groovy:
```diff
@@ -0,0 +1,264 @@
+        SET
+            TITLE =                 :title,
+            DESCRIPTION =           :description,
+            PLACE_ID =              (SELECT PLACE_ID FROM PLACES
```
Same as above

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -0,0 +1,239 @@
+        }
+
+        //get custom fields and filters ready for DAO
+        String customFieldData = JsonOutput.toJson(newEvent.customFields)
```
What if the data contains malformed json?

jaredkosanovic replied:
Malformed json is caught before the method is called, handled by this annotation for POST and PATCH: 
`@Valid ResultObject newResultObject`

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -0,0 +1,239 @@
+        }
+
+        eventsDAO.createEvent(newResourceObject.id, newEvent, filterData, customFieldData)
+
```
Quite a bit of what this method does is validation. Can the validation be moved to another method that returns a list of errors? This method could then generate an array of errors to return to the user.

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -0,0 +1,239 @@
+        resultObject.data = []
+
+        events.each {
+            println(it.id)
```
Extra println statement. You could change it to a log.debug statement so that it gets printed based on the debug level of the API.

jaredkosanovic replied:
This was something I forgot to remove.

Commit:
added check for null DELETED_AT value

Commit:
removed unneeded println

Commit:
added data validation method for post

Commit:
clean up, add some checks

Commit:
small bug with instance id

Commit:
let patch use data validation method

Commit:
add template bad request error

Commit:
clean up some data validation things

Commit:
indent

Commit:
make it jenkins friendly

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -221,7 +199,75 @@ class EventsResource extends Resource {
+/**
+ * Helper function for data validation. Returns list of applicable errors.
+ */
+    private List<Error> getErrors(ResultObject resultObject,
```
This validation method would be a good candidate for unit tests.

jaredkosanovic replied:
Good point. I made a note in CO-689.

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -221,7 +199,75 @@ class EventsResource extends Resource {
+
+        if (pathID) {
+            if (!eventsDAO.getById(pathID)) {
+                errors.add(new Error(
```
The skeleton should have a method that can returns a notFound error object. There's the notFound() method but that returns a ResponseBuilder object. Can you create a separate ticket to make that enhancement and update this to just use the skeleton?

jaredkosanovic replied:
Created CO-706

floppyears commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -221,7 +199,75 @@ class EventsResource extends Resource {
+            }
+            if (resultObject.data.id != pathID) {
+                Error mismatchID = ErrorMessages.badRequest
+                mismatchID.developerMessage = "ID in JSON body must match ID in path parameter"
```
This string should also be a constant like the other error messages.

Commit:
moved message to a variable

also deleted unused message variable

Commit:
added simple heath check

Commit:
added needed import

Commit:
return 404 for getByID

Commit:
add methods to check data

these methods check departments, groups, and places. Also made a slight change to the departments query to handle a new column in the db

Commit:
check departments, groups, and places

also cleaned the data validation method up a little bit

Commit:
added customfield and filter checking

Commit:
#hashtagcheck

Commit:
added example postman call

Commit:
comments can be nice

Commit:
small comment
